src/OptimalControlUtilities/include/BipedalLocomotionControllers/OptimalControlUtilities/ControlProblemElements.h

```diff
@@ -0,0 +1,745 @@
+     * @param velocity desired  angular velocity
+     * @param rotation desired rotation matrix
+     * @throw std::runtime_error if the Type is different from POSE and ORIENTATION
+     */
+    void setDesiredTrajectory(const iDynTree::Vector3& acceleration,
+                              const iDynTree::Vector3& velocity,
+                              const iDynTree::Rotation& rotation);
+
+    /**
+     * Set the desired trajectory
+     * @param acceleration desired spatial acceleration
+     * @param velocity angular desired spatial velocity
+     * @param transform desired homogeneous transform
+     * @throw std::runtime_error if the Type is different from POSE
+     */
+    void setDesiredTrajectory(const iDynTree::SpatialAcc& acceleration,
+                              const iDynTree::Twist& velocity,
+                              const iDynTree::Transform& transform);
+
+    /**
+     * Set the desired trajectory
+     * @param acceleration desired acceleration
+     * @param velocity angular desired velocity
+     * @param transform desired position
+     * @throw std::runtime_error if the Type is different from ONE_DIMENSION
+     */
+    void setDesiredTrajectory(const double& acceleration,
+                              const double& velocity,
+                              const double& position);
+
+    /**
+     * Set the linear PID gains
+     * @param kp proportional gain
+     * @param kd derivative gain
+     * @throw std::runtime_error if the Type is different from POSE and POSITION
+     */
+    void setLinearPIDGains(const iDynTree::Vector3& kp, const iDynTree::Vector3& kd);
+
+    /**
+     * Set the orientation PID gains
+     * @param c0 c0 gain
+     * @param c1 derivative gain
+     * @param c2 proportional gain
+     * @throw std::runtime_error if the Type is different from ORIENTATION and POSITION
+     */
+    void setOrientationPIDGains(const double& c0, const double& c1, const double& c2);
+
+    /**
+     * Set the linear PID gains
+     * @param kp proportional gain
+     * @param kd derivative gain
+     * @throw std::runtime_error if the Type is different from ONE_DIMENSION
+     */
+    void setOneDegreePIDGains(const double& kp, const double& kd);
+
+    /**
+     * Get (and compute) the element matrix
+     * @return the element matrix
+     */
+    virtual const iDynTree::MatrixDynSize& getA() final;
+
+    /**
+     * Get (and compute) the element vector
+     * @return the element vector
+     */
+    virtual const iDynTree::VectorDynSize& getB() final;
+
+    /**
+     * Set if the link is in contact with the environment
+     * @param isInContact true if the link associated to the frame is in contact with the
+     * environment
+     */
+    void isInContact(bool isInContact);
+};
+
+/**
+ * SystemDynamicsElement describes the System Dynamics that will be embedded as
+ * equality constraint or cost function.
+ */
+class SystemDynamicsElement : public CostFunctionOrEqualityConstraintElement
```

Comment:
It is not a generic system. What about ``RobotDynamicsElement`` or ``MultiBodyDynamicsElement``?

Reply:
Probably something that remembers that it is a floating base system could be used. What do you think about `FloatingBaseMultiBodyDynamicsElement`. Is it too long?

Reply:
I agree it is pretty long, but I think it's fine

Reply:
Done in d8fb355ead68cf1d1dc809d3e96ad6132c73e364